File: Project Files/JetPro_Project.m
%% AE 4451 Jet & Rocket Propulsion Project
%10/26/2018
%Authors: Loren Isakson, Brandon Campanile, Matthew Yates


%% Main Function

function out = JetPro_Project(T, eType, Nmix, Ta, Pa, Pf, M, Prf, Prc, Prb, Prab, Prnm, beta, b, f, fab, Tomax, Tmax_ab, MW, eff, y, HVf)
% Ta = ambient temperature [Kelvin]; Pa = ambient pressure [kPa]; Pf = fuel storage pressure [kPa];
% M = flight mach number Prc = compressor stagnation pressure ratio; Prf = fan stagnation pressure ratio;
% Prb = burner stagnation pressure ratio; beta = bypass ratio; b = bleed ratio; f = main burner fuel-air ratio;
% fab = afterburner fuel-air ratio; Nmix = nozzle mixing boolean; Tmax = burner max stagnation temp [K];
% Tmax_ab = max stagnation temp afternburner [K]; Prab = stagnation pressure ratio afterburner; MW = list of all
% molecular weights; y = list of all specific heat ratios; eff = list of component efficiencies
% eType = 1 for turbofan 0 for turbojet; Nmix = nozzle mixing?; T = 1 for final run, 0 for optimization run 

ynm=y(8);
R = 8314;
u = M*sqrt(y(1)*Ta*R/MW(1)); %vaircraft velocity

% Run components based on engine cycles
if strcmp(eType,'Turbojet')
    beta=0;
    [To1, Po1] = diffuser(Ta, Pa, M, y(1), eff(1));
    [To3, Po3, wc_ma] = compressor(To1, Po1, Prc, MW(3), y(3), eff(3));
    [To4, Po4, wp_ma, fmax, f] = burner(To3, Po3, Prb, b, f, MW(4), y(4), eff(4), eff(11), HVf, Tomax, Pf, fab);
    [To5_1, Po5_1] = turbine(To4, Po4, wc_ma, wp_ma, b, f, MW(5), y(5), eff(5));
    [To5_m, Po5_m] = turbineMixer(To5_1, Po5_1, To3, f, b, MW(6), y(6));    
    Po1=Po1/1000;
    Po2='NA';
    Po3=Po3/1000;
    wf_ma='NA';
    To2='NA';
    Po5_2='NA';
    To5_2='NA';
    Po7='NA';
    To7='NA';
    Tef='NA';
    Pef='NA';
    uef='NA';
    Tec='NA';
    Pec='NA';
    uec='NA';
    Po4=Po4/1000;
    wc_ma=wc_ma/1000;
    wp_ma=wp_ma/1000;
    Po5_1=Po5_1/1000;
    if fab>0 % with afterburner
        [To6, Po6, fabmax, fab] = afterburner(Po5_m, To5_m, Prab, f, fab, fmax, MW(8), y(8), eff(7), HVf, Tmax_ab);
        [Te, Pe, ue] = coreNozzle(To6, Po6, Pa, MW(9), y(9), eff(8));
        [ST, TSFC, effth, effp, effo] = performance(f, fab, ue, 0, u, beta, Pa, M, HVf);
        Po5_m=Po5_m/1000;
        Po6=Po6/1000;
        Pe=Pe/1000;
    else % without afterburner
        [Te, Pe, ue] = coreNozzle(To5_m, Po5_m, Pa, MW(9), y(9), eff(8));
        [ST, TSFC, effth, effp, effo] = performance(f, fab, ue, 0, u, beta, Pa, M, HVf);
        Po5_m=Po5_m/1000;
        Po6='NA';
        To6='NA';
        Pe=Pe/1000;
        fabmax='NA';
    end
elseif strcmp(eType,'Turbofan')
    [To1, Po1] = diffuser(Ta, Pa, M, y(1), eff(1));
    [To2, Po2, wf_ma] = fan(To1, Po1, Prf, beta, MW(2), y(2), eff(2));
    [To3, Po3, wc_ma] = compressor(To2, Po2, Prc, MW(3), y(3), eff(3));
    [To4, Po4, wp_ma, fmax, f] = burner(To3, Po3, Prb, b, f, MW(4), y(4), eff(4), eff(11), HVf, Tomax, Pf, fab);
    [To5_1, Po5_1] = turbine(To4, Po4, wc_ma, wp_ma, b, f, MW(5), y(5), eff(5));
    [To5_m, Po5_m] = turbineMixer(To5_1, Po5_1, To3, f, b, MW(6), y(6));
    [To5_2, Po5_2] = fanTurbine(To5_m, Po5_m, wf_ma, f, MW(7), y(7), eff(6));
    Po1=Po1/1000;
    Po3=Po3/1000;
    Po4=Po4/1000;
    wf_ma=wf_ma/1000;
    wc_ma=wc_ma/1000;
    wp_ma=wp_ma/1000;
    Po5_1=Po5_1/1000;
    Po5_m=Po5_m/1000;
    if fab>0 && Nmix % with afterburner and nozzle mixing 
        ue='NA';
        Tef='NA';
        Pef='NA';
        uef='NA';
        Te='NA';
        Pe='NA';
        ue='NA';
        [To6, Po6, fabmax, fab] = afterburner(Po5_2, To5_2, Prab, f, fab, fmax, MW(8), y(8), eff(7), HVf, Tmax_ab);
        [To7, Po7, ynm] = nozzleMixer(To6, Po6, beta, f, fab, Po2, To2, Prnm);
        [Tec, Pec, uec] = combinedNozzle(To7, Po7, Pa, MW(12), y(11), eff(10));
        [ST, TSFC, effth, effp, effo] = performance(f, fab, uec, uec, u, beta, Pa, M, HVf);
        Po7=Po7/1000;
        Po6=Po6/1000;
        Po5_2=Po5_2/1000;
        Pec=Pec/1000;
        Po2=Po2/1000;
    elseif fab>0 && ~Nmix % with afterburner, no nozzle mixing
        To7='NA';
        Po7='NA';
        ynm='NA';
        uec='NA';
        Tec='NA';
        Pec='NA';
        [To6, Po6, fabmax, fab] = afterburner(Po5_2, To5_2, Prab, f, fab, fmax, MW(8), y(8), eff(7), HVf, Tmax_ab);
        [Tef, Pef, uef] = fanNozzle(To2, Po2, Pa, MW(10), y(10), eff(9));
        [Te, Pe, ue] = coreNozzle(To6, Po6, Pa, MW(9), y(9), eff(8));
        [ST, TSFC, effth, effp, effo] = performance(f, fab, ue, uef, u, beta, Pa, M, HVf);
        Pef = Pef/1000;
        Pe=Pe/1000;
        Po5_2=Po5_2/1000;
        Po6=Po6/1000;
        Po2=Po2/1000;
    elseif fab==0 && Nmix % with nozzle mixing, no afterburner
        To6='NA';
        Po6='NA';
        ue='NA';
        uef='NA';
        Pe='NA';
        Te='NA';
        Tef='NA';
        Pef='NA';
        fabmax='NA';
        [To7, Po7, ynm] = nozzleMixer(To5_2, Po5_2, beta, f, fab, Po2, To2, Prnm);
        [Tec, Pec, uec] = combinedNozzle(To7, Po7, Pa, MW(12), y(11), eff(10));
        [ST, TSFC, effth, effp, effo] = performance(f, fab, uec, uec, u, beta, Pa, M, HVf);
        Po7=Po7/1000;
        Pec=Pec/1000;
        Po5_2=Po5_2/1000;
        Po7=Po7/1000;
        Po2=Po2/1000;
    else % no nozzle mixing or afterburner
        To6='NA';
        Po6='NA';
        To7='NA';
        Po7='NA';
        uec='NA';
        Tec='NA';
        Pec='NA';
        ynm='NA';
        fabmax='NA';
        [Tef, Pef, uef] = fanNozzle(To2, Po2, Pa, MW(10), y(10), eff(9));
        [Te, Pe, ue] = coreNozzle(To5_2, Po5_2, Pa, MW(9), y(9), eff(8));
        [ST, TSFC, effth, effp, effo] = performance(f, fab, ue, uef, u, beta, Pa, M, HVf);
        Pef = Pef/1000;
        Pe = Pe/1000;
        Po5_2 = Po5_2/1000;
        Po2=Po2/1000;
        Pef=Pef/1000;
        Pe=Pe/1000;
    end
else %Ramjet
    beta=0;
    [To1, Po1] = diffuser(Ta, Pa, M, y(1), eff(1));
    To2='NA';
    Po2='NA';
    To3='NA';
    Po3='NA';
    wc_ma='NA';
    [To4, Po4, wp_ma, fmax, f] = burner(To1, Po1, Prb, b, f, MW(4), y(4), eff(4), eff(11), HVf, Tomax, Pf, fab);
    wp_ma=wp_ma/1000;
    To5_1='NA';
    Po5_1='NA';
    To5_m='NA';
    Po5_m='NA';
    To5_2='NA';
    Po5_2='NA';
    wf_ma='NA';
    Tec='NA';
    Pec='NA';
    uec='NA';
    Po1=Po1/1000;
    if fab>0 % with afterburner
        To7 = 'NA';
        Po7='NA';
        Tef='NA';
        Pef='NA';
        uef='NA';
        [To6, Po6, fabmax, fab] = afterburner(Po4, To4, Prab, f, fab, fmax, MW(8), y(8), eff(7), HVf, Tmax_ab);
        [Te, Pe, ue] = coreNozzle(To6, Po6, Pa, MW(9), y(9), eff(8));
        [ST, TSFC, effth, effp, effo] = performance(f, fab, ue, 0, u, beta, Pa, M, HVf);
        Po4=Po4/1000;
        Po6=Po6/1000;
        Pe=Pe/1000;
    else % without afterburner
        To6='NA';
        Po6='NA';
        To7='NA';
        Po7='NA';
        Tef='NA';
        Pef='NA';
        uef='NA';
        fabmax='NA';
        [Te, Pe, ue] = coreNozzle(To4, Po4, Pa, MW(9), y(9), eff(8));
        [ST, TSFC, effth, effp, effo] = performance(f, fab, ue, 0, u, beta, Pa, M, HVf);
        Po4=Po4/1000;
        Pe=Pe/1000;
    end
end
if T
    % Generate Table for outputs
    titletop = [{'Inputs'}, cell(1,11)];
    inputs = {'Ta(K)','Pa(kPa)','M','Prc','Prf','beta','b','f','fab';
        Ta,Pa/1000,M,Prc,Prf,beta,b,f,fab};
    inputs = [inputs cell(2,3)];
    titlebot = [{'Outputs'}, cell(1,11)];
    output1 = {'To1(K)','Po1(kPa)','To2(K)','Po2(kPa)','To3(K)','Po3(kPa)','To4(K)','Po4(kPa)','To5.1','Po5.1(kPa)','To5.m(K)','Po5.m(kPa)';
        To1,Po1,To2,Po2,To3,Po3,To4,Po4,To5_1,Po5_1,To5_m,Po5_m};
    output2 = {'To5.2(K)','Po5.2(kPa)','To6(K)','Po6(kPa)','To7(K)','Po7(kPa)','Te(K)','Pe(kPa)','ue(m/s)','Tef(K)','Pef(kPa)','uef(m/s)';
        To5_2,Po5_2,To6,Po6,To7,Po7,Te,Pe,ue,Tef,Pef,uef};
    
    perform1 = {'Tec(K)','Pec(kPa)','uec(m/s)','ynm','ST(kNs/kg)','TSFC(kg/kNs)','nth(%)','np(%)','no(%)','Wc(kJ/kg)','Wp(kJ/kg)','Wft(kJ/kg)';
        Tec, Pec, uec, ynm,ST/1000,TSFC*1000,effth*100,effp*100,effo*100,wc_ma,wp_ma,wf_ma};
    perform2 = {'fmax','fmaxab','','','','','','','','','','';fmax,fabmax,'','','','','','','','','',''};
    warning('off','MATLAB:xlswrite:AddSheet')
    xlswrite('Results.xlsx',[titletop;inputs;cell(2,12);titlebot;output1;cell(1,12);output2;cell(1,12);perform1;cell(1,12);perform2]);
end

if strcmp(eType,'Turbojet')
    out = [ST, TSFC, To3, To5_m, f, fab];
elseif strcmp(eType,'Turbofan')
    out = [ST, TSFC, To3, To5_2, f, fab];
else
    out = [ST, TSFC, To1, To4, f, fab];
end
end

%% Component Functions

function [To1, Po1] = diffuser(Ta, Pa, M, yd, nd)
%static ambient temp, press, mach number, gamma, adiabatic efficiency

if M<=1
    rd=1;
else
    rd=1-.075*(M-1)^1.35;
end
To1 = Ta.*(1+0.5.*(yd-1).*M.^2);
Po1 = (Pa.*(1+nd.*(To1/Ta - 1)).^(yd/(yd-1)))*rd;
end

function [To2, Po2, wf_ma] = fan(To1, Po1, Prf, beta, MW, yf, nf)
%stagnation temp, press, pressure ratio, polytropic efficiency
R = 8314;
Cp = yf*(R/MW)/(yf-1);
To2 = To1.*(Prf).^((yf-1)/yf/nf);
Po2 = Po1.*Prf;
wf_ma = Cp*(1+beta)*(To2-To1); % work done by the fan on the fluid
end

function [To3, Po3, wc_ma] = compressor(To2, Po2, Prc, MW, yc, nc)
R = 8314;
Cp = yc*(R/MW)/(yc-1);

Po3 = Po2*Prc;
To3 = To2*(Prc).^((yc-1)/yc/nc);
wc_ma = Cp*(To3-To2);
end

function [To4, Po4, wp_ma, fmax, f] = burner(To3, Po3, Prb, b, f, MW, yb, nb, nfp, HVf, Tomax, Pf, fab)
Cb = 700; %kelvin
R = 8314;
bmax = 0.12;
deltaP_inject = 550*10^3;
f_density = 780;

Tmax = Tomax + Cb*(b/bmax)^0.5;
Cp = yb*(R/MW)/(yb-1);
fmax = (1-b)*(1-To3/Tmax)/((nb*HVf/Cp/Tmax)-1);

To4 = (1/(1-b+f))*((1-b)*To3+nb*HVf*f/Cp);
Po4 = Po3*Prb;

%Fuel Pump
wp_ma = (f+fab)*(Po3+deltaP_inject-Pf)/nfp/f_density;
end

function [To5_1, Po5_1] = turbine(To4, Po4, wc_ma, wp_ma, b, f, MW, yt, nt)
R = 8314;
Cp = yt*(R/MW)/(yt-1);
To5_1 = To4 - (wc_ma + wp_ma)/(Cp*(1+f-b));
TR = To5_1/To4;
Po5_1 = Po4*(TR^(1/nt))^(yt/(yt-1));
end

function [To5_m, Po5_m] = turbineMixer(To5_1, Po5_1, To3, f, b, MW, ytm)
%R = 8314;
%Cp = ytm*(R/MW)/(ytm-1);
To5_m = (b*To3 + (1+f-b)*To5_1)/(1+f);
Po5_m = Po5_1*((To5_m/To5_1)^(ytm/(ytm-1)))*(To5_1/To3)^(ytm*b/((ytm-1)*(1+f)));
end

function [To5_2, Po5_2] = fanTurbine(To5_m, Po5_m, wf_ma, f, MW, yft, nft)
R = 8314;
Cp = yft*(R/MW)/(yft-1);
To5_2 = To5_m - wf_ma/(Cp*(1+f));
TR = To5_2/To5_m;
Po5_2 = Po5_m*(TR^(1/nft))^(yft/(yft-1));
end

function [To6, Po6, fabmax, fab] = afterburner(Po5_2, To5_2, Prab, f, fab, fmax, MW, yab, nab, HVf, Tmax_ab)
R = 8314;
cpab = yab*(R/MW)/(yab-1);
fabmax = (1+fmax)*((Tmax_ab/To5_2)-1)/((nab*HVf/cpab/To5_2)-(Tmax_ab/To5_2));

Po6 = Po5_2*Prab;
To6 = (1/(1+f+fab))*((1+f)*To5_2+nab*HVf*fab/cpab);
end

function [Te, Pa, ue] = coreNozzle(To6, Po6, Pa, MW, yn, nn)
R = 8314;
Cp = yn*(R/MW)/(yn-1);
Te = To6*(1-nn*(1-(Pa/Po6)^((yn-1)/yn)));
ue = sqrt(2*Cp*(To6-Te));
if ~isreal(ue)
    ue=0;
end
end

function [Tef, Pa, uef] = fanNozzle(To2, Po2, Pa, MW, yfn, nfn)
R = 8314;
Cp = yfn*(R/MW)/(yfn-1);
Tef = To2*(1-nfn*(1-(Pa/Po2)^((yfn-1)/yfn)));
uef = sqrt(2*Cp*(To2-Tef));
end

function [To7, Po7, gamma] = nozzleMixer(To6, Po6, beta, f, fab, Po2, To2, Prnm)
To7 = (beta*To2+(1+f+fab)*To6)/(1+beta+f+fab);
gamma = 1.44 - (1.39*10^-4)*To7 + (3.57*10^-8)*To7^2;
Po7 = Po6*Prnm*((Po2/Po6)^(beta/(1+beta+f+fab)))*((To7/To6)^(gamma/(gamma-1)))*((To6/To2)^((gamma*beta)/(gamma-1)/(1+beta+f+fab)));
end

function [Tec, Pa, uec] = combinedNozzle(To7, Po7, Pa, MW, ycn, ncn)
R = 8314;
Cp = ycn*(R/MW)/(ycn-1);
Tec = To7*(1-ncn*(1-(Pa/Po7)^((ycn-1)/ycn)));
uec = sqrt(2*Cp*(To7-Tec));
if ~isreal(uec)
    uec=0;
end
end


%% Specific Thrust, Specific Fuel Consumption, Efficiencies

function [ST, TSFC, effth, effp, effo] = performance(f, fab, ue, uef, u, beta, Pa, M, HVf)
Patm = 101.3*10^3;
CB = 245;
deltaD = CB*M^2*Pa/Patm*beta^1.5;
ST = (1+f+fab)*ue+beta*uef-(beta+1)*u - deltaD;
TSFC = (f+fab)/ST;
effth = (beta*uef^2+(1+f+fab)*ue^2-(beta+1)*u^2)/(2*(f+fab)*HVf);
effp = 2*ST*u/(beta*uef^2+(1+f+fab)*ue^2-(beta+1)*u^2);
effo = effth*effp;
end

File: Project Files/JetPro_Project.py
Rbar = 8314
Ta = 220
Pa = 10*10**3
Patm = 101325
M = 1.50
PRc = 30
PRf = 1.2
B = 2
b = 0.1
f = 0.018
fab = 0.010
nadd = 0.92
yd = 1.4
MW = 28.8
yf = 1.4
npf = 0.90
npc = 0.90
yc = 1.38
nb = 0.99
yb = 1.33
delH = 45*10**6
PRb = 0.98
Tmaxo = 1300
Cb1 = 700
bmax = 0.12
yt = 1.33
npt = 0.92
delp = 550*10**3
rhof = 780
nadp = 0.35
ytm = 1.34
npft = 0.92
yft = 1.33
nab = 0.96
yab = 1.32
PRab = 0.97
nadn = 0.95
yn = 1.35
nadfn = 0.97
yfn = 1.4
PRnm = 0.8
nadcn = 0.95
ycn = 1.37
u = M*(1.4*Rbar*Ta/MW)**0.5
CB1 = 245
deld = CB1*(M**2)*(Pa/Patm)*(B**1.5)

def diffuser(Ta, Pa, M, nad, y):
    if M<=1:
        rd = 1
    elif M>1 and M<5:
        rd = 1-0.075*(M-1)**1.35
    Po1 = Pa*rd*(1+nad*0.5*(y-1)*M**2)**(y/(y-1))
    To1 = Ta*(1+0.5*(y-1)*M**2)
    return Po1, To1

def fan(To1, Po1, npf, y, MW, PRf):
    cpf = y*(Rbar/MW)/(y-1)
    To2 = To1*PRf**((y-1)/(y*npf))
    Po2 = Po1*PRf
    Wf_ma = (1+B)*cpf*(To2-To1)
    return Po2, To2, Wf_ma

def compressor(To2, Po2, PRc, npc, y, MW):
    cpc = y*(Rbar/MW)/(y-1)
    To3 = To2*PRc**((y-1)/(y*npc))
    Po3 = Po2*PRc
    Wc_ma = cpc*(To3-To2)
    return Po3, To3, Wc_ma

def burner(To3, Po3, PRb, y, MW, nb, delH, Tmaxo, Cb1, b, bmax):
    cpb = y*(Rbar/MW)/(y-1)
    To4 = (1/(1-b+f))*((1-b)*To3+nb*delH*f/cpb)
    Po4 = Po3*PRb
    Tmax = Tmaxo +Cb1*(b/bmax)**0.5
    #fmax = (1-To3/To4)/(delH*nb/cpb/To4-1) #needs improvement
    return Po4, To4, Tmax

def fuel_pump(delp, rhof, nadp, f):
    Wp_ma = f*delp/rhof/nadp
    return Wp_ma

def turbine(To4, Po4, Wc, Wp, y, npt, MW):
    cpt = y*(Rbar/MW)/(y-1)
    To5_1 = To4 - Wc/(cpt*(1+f-b))-Wp/(cpt*(1+f-b))
    TR = To5_1/To4
    Po5_1 = Po4*(TR**(1/npt))**(y/(y-1))
    return Po5_1, To5_1

def tmixer(To3, To5_1, Po5_1, f, b, y):
    To5_m = (b*To3+(1+f-b)*To5_1)/(1+f)
    #Po5_m = Po5_1
    #Po5_m = Po5_1*((To5_m/To5_1)**(y/(y-1)))*((To5_1/To3)**(y*b/((y-1)*(1+f-b))))
    Po5_m = Po5_1*((To5_m/To5_1)**(y/(y-1)))*((To5_1/To3)**(y*b/((y-1)*(1+f))))
    return Po5_m, To5_m

def fan_turbine(To5_m, Po5_m, npft, y, Wf, f, MW):
    cpft = y*(Rbar/MW)/(y-1)
    To5_2 = To5_m - Wf/(cpft*(1+f))
    TR = To5_2/To5_m
    Po5_2 = Po5_m*(TR**(1/npft))**(y/(y-1))
    return Po5_2, To5_2

def afterburner(To5_2, Po5_2, PRab, y, MW, fab, f, delH, nab):
    cpab = y*(Rbar/MW)/(y-1)
    if fab>0:
        PR = PRab
    else:
        PR = 1
    Po6 = Po5_2*PR
    To6 = (1/(1+f+fab))*((1+f)*To5_2+nab*delH*fab/cpab)
    return Po6, To6

def nozzle(To6, Po6, Pa, nadn, y, MW):
    cpn = y*(Rbar/MW)/(y-1)
    Te = To6*(1-nadn*(1-(Pa/Po6)**((y-1)/y)))
    Pe = Pa
    ue = (2*cpn*(To6-Te))**0.5
    return Pe, Te, ue

def fan_nozzle(To2, Po2, Pa, nadfn, y, MW):
    cpfn = y*(Rbar/MW)/(y-1)
    Tef = To2*(1-nadfn*(1-(Pa/Po2)**((y-1)/y)))
    Pef = Pa
    uef = (2*cpfn*(To2-Tef))**0.5
    return Pef, Tef, uef

def nozzle_mixer(To6, Po6, To2, Po2, PRnm, B, f, fab):
    To7 = (B*To2+(1+f+fab)*To6)/(1+B+f+fab)
    y = 1.44-(1.39*10**-4)*To7+(3.57*10**-8)*To7**2
    Po7 = Po6*PRnm*((Po2/Po6)**(B/(1+B+f+fab)))*((To7/To6)**(y/(y-1)))*((To6/To2)**(y*B/((y-1)*(1+B+f+fab))))
    return Po7, To7, y

def combined_nozzle(To7, Po7, Pa, y, MW, nadcn):
    cpcn = y*(Rbar/MW)/(y-1)
    Tec = To7*(1-nadcn*(1-(Pa/Po7)**((y-1)/y)))
    Pec = Pa
    uec = (2*cpcn*(To7-Tec))**0.5
    return Pec, Tec, uec

def st(ue, uef, u, f, fab, B, deld):
    ST = (1+f+fab)*ue+B*uef-(1+B)*u-deld
    return ST

def stcn(uec, u, f, fab, B, deld):
    STCN = (1+B+f+fab)*uec-(1+B)*u-deld
    return STCN

def tsfc(ST, f, fab):
    TSFC = (f+fab)/ST
    return TSFC

def prop_eff(ST, ue, uef, u, f, fab, B):
    KE = (1+f+fab)*0.5*ue**2+B*0.5*uef**2-(1+B)*0.5*u**2
    np = ST*u/KE
    return np, KE

def prop_eff_cn(STCN, uec, u, f, fab, B):
    KECN = (1+B+f+fab)*0.5*uec**2-(1+B)*0.5*u**2
    npcn = STCN*u/KECN
    return npcn, KECN

def therm_eff(KE, f, fab, delH):
    nth = KE/((f+fab)*delH)
    return nth

Po1, To1 = diffuser(Ta, Pa, M, nadd, yd)
Po2, To2, Wf_ma = fan(To1, Po1, npf, yf, MW, PRf)
Po3, To3, Wc_ma = compressor(To2, Po2, PRc, npc, yc, MW)
Po4, To4, Tmax = burner(To3, Po3, PRb, yb, MW, nb, delH, Tmaxo, Cb1, b, bmax)
Wp_ma = fuel_pump(delp, rhof, nadp, f)
Po5_1, To5_1 = turbine(To4 , Po4, Wc_ma, Wp_ma, yt, npt, MW)
Po5_m, To5_m = tmixer(To3, To5_1, Po5_1, f, b, ytm)
Po5_2, To5_2 = fan_turbine(To5_m, Po5_m, npft, yft, Wf_ma, f, MW)
Po6, To6 = afterburner(To5_2, Po5_2, PRab, yab, MW, fab, f, delH, nab)
Pe, Te, ue = nozzle(To6, Po6, Pa, nadn, yn, MW)
Pef, Tef, uef = fan_nozzle(To2, Po2, Pa, nadfn, yfn, MW)
Po7, To7, ynm = nozzle_mixer(To6, Po6, To2, Po2, PRnm, B, f, fab)
Pec, Tec, uec = combined_nozzle(To7, Po7, Pa, ycn, MW, nadcn)
ST = st(ue, uef, u, f, fab, B, deld)
STCN = stcn(uec, u, f, fab, B, deld)
TSFC = tsfc(ST, f, fab)
TSFCCN = tsfc(STCN, f, fab)
np, KE = prop_eff(ST, ue, uef, u, f, fab, B)
npcn, KECN = prop_eff_cn(STCN, uec, u, f, fab, B)
nth = therm_eff(KE, f, fab, delH)
nthcn = therm_eff(KECN, f, fab, delH)
no = np*nth
nocn = npcn*nthcn